## ✈️ *TravelTide* - Customer Segmentation Project
---
### Data Understanding and Preparation - SQL Exploration

> ⚠️ **Execution Notice – Notebook 02**
>
> This notebook performs **large-scale SQL exploration and validation**
> on the raw TravelTide database using **Spark + JDBC**,
> including tables with more than **5 million records** (e.g. `sessions`).
>
> The full exploratory SQL workflow is **designed to run in Databricks**
> and is **not intended to be executed end-to-end in Colab**.
>
> **Important:**  
> When running this notebook outside Databricks (e.g. in Colab),
> **do not run the entire notebook**.
>
> Only the two cells under the section  
> **“Session-Level Dataset Materialization”**
> are expected to be executed.
>
> All other SQL cells serve a **documentation and validation purpose**:
> they record the analytical reasoning used to understand the data model,
> verify lifecycle logic, and justify cohort and filtering decisions.
>
> The materialized session-level dataset produced in this step
> is reused by all downstream notebooks (**Notebook 03 onward**).


### **Session-Level Dataset Materialization**

> 🔹 **Execution Scope**  
> The following two cells represent the **only executable steps**
> required to reproduce this notebook outside Databricks.
> They materialize the session-level dataset used by all downstream analyses.


In [7]:
# Install required database and ORM libraries
# psycopg2-binary: PostgreSQL driver
# sqlalchemy: database connection and query interface
!pip -q install psycopg2-binary sqlalchemy

# Core data handling
import pandas as pd

# Database connection utilities
from sqlalchemy import create_engine

# File system and path management
from pathlib import Path
import os

# Mount Google Drive to persist data and outputs
from google.colab import drive
drive.mount('/content/drive')


# Project directory setup
# -------------------------------------------------------------------

# Root project directory on Google Drive
PROJECT_DIR = Path("/content/drive/MyDrive/Masterschool/TravelTide")

# Raw data directory (created if it does not exist)
(DATA_RAW := PROJECT_DIR / "data" / "raw").mkdir(parents=True, exist_ok=True)


# Database connection parameters
# -------------------------------------------------------------------

# Public PostgreSQL database
DB_HOST = "ep-noisy-flower-846766.us-east-2.aws.neon.tech"
DB_NAME = "TravelTide"
DB_USER = "Test"
DB_PW   = "bQNxVzJL4g6u"

# Create SQLAlchemy engine for PostgreSQL
engine = create_engine(
    f"postgresql+psycopg2://{DB_USER}:{DB_PW}@{DB_HOST}/{DB_NAME}"
)


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [6]:
query = """
WITH
  -- 1. Sessions after cutoff data
  filtered_sessions AS (
    SELECT *
    FROM sessions
    WHERE session_start > '2023-01-04'
  ),

  -- 2. Users with sufficient activity
  filtered_users AS (
    SELECT
      user_id,
      COUNT(*) AS session_count
    FROM filtered_sessions
    GROUP BY user_id
    HAVING COUNT(*) > 7
  ),

  -- 3. Session-level table for filtered users
  session_level AS (
    SELECT
      -- SESSION FIELDS -------------
      s.session_id,
      s.user_id,
      s.trip_id,
      s.session_start,
      s.session_end,
      s.page_clicks,
      s.flight_discount,
      s.flight_discount_amount,
      s.flight_booked,
      s.hotel_discount,
      s.hotel_discount_amount,
      s.hotel_booked,
      s.cancellation,

      -- USERS FIELDS -------------
      u.birthdate,
      u.gender,
      u.married,
      u.has_children,
      u.home_country,
      u.home_city,
      u.home_airport,
      u.home_airport_lat,
      u.home_airport_lon,
      u.sign_up_date,

      -- FLIGHT FIELDS -------------
      f.origin_airport,
      f.destination,
      f.destination_airport,
      f.seats,
      f.return_flight_booked,
      f.departure_time,
      f.return_time,
      f.checked_bags,
      f.trip_airline,
      f.destination_airport_lat,
      f.destination_airport_lon,
      f.base_fare_usd,

      -- HOTEL FIELDS -------------
      h.hotel_name,
      h.nights,
      h.rooms,
      h.check_in_time,
      h.check_out_time,
      h.hotel_per_room_usd

    FROM filtered_sessions s
    INNER JOIN users u ON s.user_id = u.user_id
    LEFT JOIN flights f ON s.trip_id = f.trip_id
    LEFT JOIN hotels h ON s.trip_id = h.trip_id
    WHERE s.user_id IN (SELECT user_id FROM filtered_users)
  ),

  -- 4. Cancellation sessions inside the window
  cancellations AS (
    SELECT *
    FROM session_level
    WHERE cancellation = TRUE
  ),

  -- 5. Retrieve booking sessions BEFORE cutoff for trips cancelled AFTER cutoff
  sessions_trip_booked_before AS (
    SELECT *
    FROM sessions
    WHERE trip_id IN (
      SELECT trip_id
      FROM cancellations
    )
    AND session_start <= '2023-01-04'  -- booking occurred before the cutoff date
  ),

  -- 6.  Build session-level rows for these earlier booking sessions
  session_level_before AS (
    SELECT
      -- SESSION FIELDS -------------
      s.session_id,
      s.user_id,
      s.trip_id,
      s.session_start,
      s.session_end,
      s.page_clicks,
      s.flight_discount,
      s.flight_discount_amount,
      s.flight_booked,
      s.hotel_discount,
      s.hotel_discount_amount,
      s.hotel_booked,
      s.cancellation,

      -- USER FIELDS -------------
      u.birthdate,
      u.gender,
      u.married,
      u.has_children,
      u.home_country,
      u.home_city,
      u.home_airport,
      u.home_airport_lat,
      u.home_airport_lon,
      u.sign_up_date,

      -- FLIGHT FIELDS -------------
      f.origin_airport,
      f.destination,
      f.destination_airport,
      f.seats,
      f.return_flight_booked,
      f.departure_time,
      f.return_time,
      f.checked_bags,
      f.trip_airline,
      f.destination_airport_lat,
      f.destination_airport_lon,
      f.base_fare_usd,

      -- HOTEL FIELDS -------------
      h.hotel_name,
      h.nights,
      h.rooms,
      h.check_in_time,
      h.check_out_time,
      h.hotel_per_room_usd

    FROM sessions_trip_booked_before s
    INNER JOIN users u ON s.user_id = u.user_id
    LEFT JOIN flights f ON s.trip_id = f.trip_id
    LEFT JOIN hotels h ON s.trip_id = h.trip_id
    WHERE s.user_id IN (SELECT user_id FROM filtered_users)
  ),

  -- 7. Combine both parts
  final_tab AS (
    SELECT * FROM session_level
    UNION ALL
    SELECT * FROM session_level_before
  )

SELECT *
FROM final_tab;
"""

# Output path for the session-level dataset
out_path = DATA_RAW / "session_level_raw.csv"

# Read the SQL query result in chunks to avoid memory issues
# (the session table contains several million records)
chunks = []

for chunk in pd.read_sql_query(query, engine, chunksize=200_000):
    chunks.append(chunk)                 # Accumulate chunks in memory
    print("Loaded chunk:", len(chunk))   # Progress log for long-running queries

# Concatenate all chunks into a single DataFrame
sl_df = pd.concat(chunks, ignore_index=True)

# Export the consolidated session-level dataset to CSV
sl_df.to_csv(out_path, index=False)

Loaded chunk: 49218


### 1. Introduction

This notebook corresponds to the **Data Understanding** phase and the early stages of **Data Preparation** within the CRISP-DM framework.

It focuses on understanding the TravelTide data model directly at the database level using SQL.

The goal is twofold:
- Explore the schema, relationships, and granularity of the data
- Construct a clean and semantically correct session-level raw dataset that will serve as the foundation for all subsequent analysis.

No statistical EDA is performed at this stage. Instead, this notebook ensures that joins, filters, and lifecycle logic are correct before any feature engineering or modeling is attempted.


**Author**: Maria Petralia  
**Project**: TravelTide - Customer Segmentation & Perk Strategy  
**Context**: MasterSchool - Data Science Program  
**Date**: Feb 2026

> **Execution note**  
> This notebook performs direct JDBC reads and exploratory SQL queries on large raw tables.  
> In particular, the `sessions` table contains ~5.4M records.
>
> As a result, **initial execution may take up to 10 minutes**, even with partitioned reads and fetch-size tuning.  
> This is expected behavior and reflects the intentional choice to validate data integrity and lifecycle logic at full granularity before downstream aggregation.

### 2. Load the tables with Spark

#### 2.1 Import Libraries

In [ ]:
# import libraries
import pandas as pd

#### 2.2 Configure PostgreSQL Connection

In [ ]:
# PostgreSQL connection URL
traveltide_url = "jdbc:postgresql://ep-noisy-flower-846766.us-east-2.aws.neon.tech/TravelTide"

# Connection properties
properties = { "user": "Test",
              "password": "bQNxVzJL4g6u",
              "driver": "org.postgresql.Driver"
              }

### 3. Load Data

#### 3.1 Load Metadata

In [ ]:
# Load metadata into Spark DataFrames
traveltide_metadata = spark.read.jdbc(
    url=traveltide_url,
    table="information_schema.tables",
    properties=properties
)

# Filter to only public tables
public_tables = traveltide_metadata.filter(
    traveltide_metadata.table_schema == "public"
)

public_tables.select("table_name").show(truncate=False)


+----------+
|table_name|
+----------+
|sessions  |
|flights   |
|users     |
|hotels    |
+----------+



**Note**: The database contains four main tables relevant for the analysis:

- **sessions** - user sessions, clicks, booking and cancellation flags
- **users** - user demographic and profile information
- **flights** - flight details associated with trips
- **hotels** - hotel details associated with trips

These were identified by querying the `information_schema.tables` filtered on the `public` schema.

In [ ]:
# JDBC read tuning
# Number of partitions for parallel JDBC reads
N_PARTITIONS = 16
# Number of rows fetched per JDBC round-trip
JDBC_FETCHSIZE = "20000"

# bounds per user_id (da users)
# Read min and max user_id from the database
# Used later as lowerBound and upperBound for partitioned JDBC reads
bounds_row = (
    spark.read.format("jdbc")
    .options(
        url=traveltide_url,
        dbtable="(SELECT MIN(user_id) AS min_id, MAX(user_id) AS max_id FROM users) t",
        fetchsize=JDBC_FETCHSIZE,
        **properties
    )
    .load()
    .first()
)

# Convert bounds to Python integers
USER_ID_MIN = int(bounds_row["min_id"])
USER_ID_MAX = int(bounds_row["max_id"])
# Log bounds for validation
print("user_id bounds:", USER_ID_MIN, USER_ID_MAX)


user_id bounds: 0 1020925


**Note**:
Tables are read using JDBC partitioning to improve execution time by enabling parallel reads and reducing overall query latency.

#### 3.2 Load Raw Tables Into Spark DataFrames

In [ ]:
# Define functions to read data from JDBC
# - partitioned by user_id
def read_jdbc_user_partitioned(table_name: str):
    return (
        spark.read.format("jdbc")
        .options(
            url=traveltide_url,
            dbtable=table_name,
            partitionColumn="user_id",
            lowerBound=str(USER_ID_MIN),
            upperBound=str(USER_ID_MAX),
            numPartitions=str(N_PARTITIONS),
            fetchsize=JDBC_FETCHSIZE,
            **properties
        )
        .load()
    )
# - simple (no partitioning)
def read_jdbc_simple(table_name: str):
    return (
        spark.read.format("jdbc")
        .options(
            url=traveltide_url,
            dbtable=table_name,
            numPartitions="2",
            fetchsize=JDBC_FETCHSIZE,
            **properties
        )
        .load()
    )

# users + sessions: partitioned
users_spark    = read_jdbc_user_partitioned("users")
sessions_spark = read_jdbc_user_partitioned("sessions")

# flights + hotels: simple
flights_spark  = read_jdbc_simple("flights")
hotels_spark   = read_jdbc_simple("hotels")


**Note**: The four core tables are loaded via JDBC into Spark DataFrames to enable SQL-based exploration at scale.  
At this stage, data is kept at full granularity to validate schema, keys, and lifecycle logic before applying any cohort filters.  
Cohort selection and session-level extraction are performed only in the final step of the notebook.

In [ ]:
# Check partition only on sessions
sessions_spark.explain(True)

== Parsed Logical Plan ==
Relation [session_id#13193,user_id#13194,trip_id#13195,session_start#13196,session_end#13197,flight_discount#13198,hotel_discount#13199,flight_discount_amount#13200,hotel_discount_amount#13201,flight_booked#13202,hotel_booked#13203,page_clicks#13204,cancellation#13205] JDBCRelation(sessions) [numPartitions=16]

== Analyzed Logical Plan ==
session_id: string, user_id: int, trip_id: string, session_start: timestamp, session_end: timestamp, flight_discount: boolean, hotel_discount: boolean, flight_discount_amount: decimal(38,18), hotel_discount_amount: decimal(38,18), flight_booked: boolean, hotel_booked: boolean, page_clicks: int, cancellation: boolean
Relation [session_id#13193,user_id#13194,trip_id#13195,session_start#13196,session_end#13197,flight_discount#13198,hotel_discount#13199,flight_discount_amount#13200,hotel_discount_amount#13201,flight_booked#13202,hotel_booked#13203,page_clicks#13204,cancellation#13205] JDBCRelation(sessions) [numPartitions=16]

==

#### 3.3 Register Temporary Views for SQL

In [ ]:
# Create temp views for each table
users_spark.createOrReplaceTempView("users")
flights_spark.createOrReplaceTempView("flights")
hotels_spark.createOrReplaceTempView("hotels")
sessions_spark.createOrReplaceTempView("sessions")

#### 3.3 Quick Data Preview

In [ ]:
%sql
-- Count the number of records in Users, Sessions, Flights and Hotels
SELECT (SELECT COUNT(user_id) FROM users) AS users_count,
(SELECT COUNT(session_id) FROM sessions) AS sessions_count,
(SELECT COUNT(trip_id) FROM flights) AS flights_count,
(SELECT COUNT(trip_id) FROM hotels) AS hotels_count;

users_count sessions_count flights_count hotels_count 1020926 5408063 1901038 1918617

**Note**: There are over 5 million records in session table and over one million users. Flights and hotels tables have almost 2 million records

In [ ]:
%sql
-- Display the first 10 rows of sessions
SELECT * FROM sessions LIMIT 10;

session_id user_id trip_id session_start session_end flight_discount hotel_discount flight_discount_amount hotel_discount_amount flight_booked hotel_booked page_clicks cancellation 190-98c09a3b41e6480092f2834685b99872 190 190-5befdd979c734b698ab47f416a4f1f9b 2023-06-10T18:36:00.000Z 2023-06-10T18:58:16.000Z false false null null true true 180 false 369-05dfdf0732034f02be3133950463c061 369 null 2023-06-10T21:38:00.000Z 2023-06-10T21:39:03.000Z false false null null false false 9 false 437-4c352bdebd434bc7b38dc5c0cf2d499a 437 null 2023-06-10T16:24:00.000Z 2023-06-10T16:24:38.000Z false false null null false false 5 false 450-4b6467e93fea4aa6afc7505e643da323 450 null 2023-06-10T17:28:00.000Z 2023-06-10T17:28:46.000Z true false 0.300000000000000000 null false false 6 false 898-eb22bdaf6c2b402681ed6ef8ae902e70 898 null 2023-06-10T19:38:00.000Z 2023-06-10T19:38:16.000Z false false null null false false 2 false 942-ff5db1d5dcab419e8ff8f7b46b7764f6 942 null 2023-06-10T17:35:00.000Z 2023-06-10T17:36:14.000Z false false null null false false 10 false 1346-11c14dcf3ead4dc6b06b576cff7cbf04 1346 1346-444f3204fcca4a649c8e3abcca044eb4 2023-06-10T08:57:00.000Z 2023-06-10T08:58:58.000Z false false null null true false 16 false 1354-d727b304ece54a24a82a52f9c3c01d5f 1354 1354-bc0aa61994094345b381b6fecb46c7f5 2023-06-10T07:05:00.000Z 2023-06-10T07:06:28.000Z false false null null true false 12 false 1679-f9c76fba94454db18df6d5259ecd0e37 1679 null 2023-06-10T19:02:00.000Z 2023-06-10T19:05:11.000Z true true 0.100000000000000000 0.050000000000000000 false false 26 false 1975-52337d053d7b46b29cf60e9ea7824c30 1975 1975-bae4bfbd12a947cca1406f113f0f37dc 2023-06-10T14:51:00.000Z 2023-06-10T14:53:26.000Z false true null 0.050000000000000000 true true 20 false

In [ ]:
%sql
-- Display the first 10 rows of users
SELECT *
FROM users
LIMIT 10;

user_id birthdate gender married has_children home_country home_city home_airport home_airport_lat home_airport_lon sign_up_date 0 1990-01-22 F false false usa minneapolis MSP 44.880000000000000000 -93.217000000000000000 2021-04-01 1 2000-11-08 M false false usa colorado springs COS 38.806000000000000000 -104.700000000000000000 2021-04-01 2 1992-09-21 M false false usa portland PDX 45.589000000000000000 -122.597000000000000000 2021-04-01 3 1996-11-27 F false false usa houston IAH 29.980000000000000000 -95.340000000000000000 2021-04-01 4 1978-01-05 M true true usa honolulu HNL 21.316000000000000000 -157.927000000000000000 2021-04-01 5 1989-10-20 F false false canada edmonton YED 53.667000000000000000 -113.467000000000000000 2021-04-01 6 1995-10-21 F false false canada edmonton YXD 53.573000000000000000 -113.521000000000000000 2021-04-01 7 1983-05-27 M false false usa fresno FAT 36.776000000000000000 -119.718000000000000000 2021-04-01 8 2003-05-20 M false true usa phoenix PHX 33.434000000000000000 -112.008000000000000000 2021-04-01 9 1977-04-13 M false true usa las vegas LAS 36.080000000000000000 -115.152000000000000000 2021-04-01

In [ ]:
%sql
-- Display the first 10 rows of fligths
SELECT *
FROM flights
LIMIT 10;

trip_id origin_airport destination destination_airport seats return_flight_booked departure_time return_time checked_bags trip_airline destination_airport_lat destination_airport_lon base_fare_usd 496435-abe4a3a4d6bc4734a3c5baacbbdd3b2e AKR charlotte CLT 1 true 2023-02-03T15:00:00.000Z 2023-02-05T15:00:00.000Z 0 JetBlue Airways 35.214000000000000000 -80.943000000000000000 118.780000000000000000 496701-690c41b3f3ad44c198bef783bc3bff85 HOU chicago ORD 1 true 2023-02-11T12:00:00.000Z 2023-02-17T12:00:00.000Z 1 United Airlines 41.786000000000000000 -87.752000000000000000 268.530000000000000000 496778-89aabe7bd3144c52811a7a61fb05a1de MSY florence FLO 2 true 2023-03-27T20:00:00.000Z 2023-04-08T20:00:00.000Z 1 Allegiant Air 34.185000000000000000 -79.724000000000000000 287.950000000000000000 496839-aca1684059e84783a43698c4bb6ea96b MDW los angeles LAX 1 true 2023-02-08T07:00:00.000Z 2023-02-15T07:00:00.000Z 1 United Airlines 33.942000000000000000 -118.408000000000000000 443.390000000000000000 496871-d3b9674337a049eb91c451a55a81860f SAT san jose SJC 1 true 2023-02-09T08:00:00.000Z 2023-02-12T08:00:00.000Z 1 American Airlines 37.362000000000000000 -121.929000000000000000 417.770000000000000000 496894-c8f1a5bc8b514a1ebd6f760a3e40becb JFK austin AUS 1 false 2023-02-03T16:00:00.000Z null 1 Southwest Airlines 30.194000000000000000 -97.670000000000000000 229.970000000000000000 496923-ffa36182213749c79022e36f2ab7eac8 HOU chicago UGN 1 true 2023-04-24T09:00:00.000Z 2023-05-18T09:00:00.000Z 0 Ryanair 41.786000000000000000 -87.752000000000000000 249.750000000000000000 496924-c51772ff630f409eb3a9ea5a8ceacce0 YZD seattle BFI 1 true 2023-02-05T07:00:00.000Z 2023-02-06T07:00:00.000Z 1 Kenmore Air 47.530000000000000000 -122.302000000000000000 594.180000000000000000 496936-8d273e20c3664ba991253d2b2c70edf3 YVR toronto YTZ 1 true 2023-02-09T11:00:00.000Z 2023-02-16T11:00:00.000Z 0 Porter Airlines 43.862000000000000000 -79.370000000000000000 614.450000000000000000 497037-eddd4a4ea0cd422e9eb6450e137e98ba FTW toronto YKZ 1 true 2023-02-06T07:00:00.000Z 2023-02-16T07:00:00.000Z 1 American Airlines 43.862000000000000000 -79.370000000000000000 343.030000000000000000

In [ ]:
%sql
-- Display the first 10 rows of hotels
SELECT *
FROM hotels
LIMIT 10;

trip_id hotel_name nights rooms check_in_time check_out_time hotel_per_room_usd 243700-21f0686574264abdba95c975edf0439e Choice Hotels - fort worth 12 1 2022-07-14T11:00:00.000Z 2022-07-26T11:00:00.000Z 144.000000000000000000 243702-5c9d79726ccd4cd680416769c9001175 NH Hotel - nashville 5 1 2022-07-18T19:32:04.785Z 2022-07-24T11:00:00.000Z 283.000000000000000000 243703-4783bb226e82421e918c9cb927b01950 Starwood - montreal 5 1 2022-07-15T11:00:00.000Z 2022-07-20T11:00:00.000Z 197.000000000000000000 243705-43773b585acd4a1c99ea53a85baa0938 InterContinental - houston 1 1 2022-07-17T12:02:57.300Z 2022-07-19T11:00:00.000Z 578.000000000000000000 243707-42fed4bbc026475aaaf333b1f2993613 NH Hotel - montreal 4 1 2022-07-21T18:43:32.925Z 2022-07-26T11:00:00.000Z 146.000000000000000000 243709-17ef94cbfd394f28aeaa80812463f626 InterContinental - phoenix 0 1 2022-07-16T13:21:21.420Z 2022-07-17T11:00:00.000Z 75.000000000000000000 243710-7be7870893914ff781bb8d698942c90e Wyndham - washington 6 1 2022-07-17T10:00:09.585Z 2022-07-23T11:00:00.000Z 107.000000000000000000 243712-8fec5eac32074d9390e48fbb3b3fc2ac Crowne Plaza - san antonio 2 1 2022-07-18T14:59:51.270Z 2022-07-21T11:00:00.000Z 154.000000000000000000 243719-15b49a84d2934c78b034075086b0cb96 Rosewood - san diego 1 1 2022-07-17T15:34:57.090Z 2022-07-19T11:00:00.000Z 99.000000000000000000 243720-aae2fe9145fa4bbf97ea9da37251b74e Hilton - los angeles 4 2 2022-07-17T18:58:03.945Z 2022-07-22T11:00:00.000Z 225.000000000000000000

**Note**: The first 10 rows of each table confirm consistency in structure. In the `sessions` table, columns like `trip_id`, discount-related fields can be `NULL`, reflecting browsing-only sessions rather than missing data.

### 4. Data Consistency Checks

#### 4.1 Data Types

In [ ]:
%sql
DESCRIBE users;

col_name data_type comment user_id int null birthdate date null gender string null married boolean null has_children boolean null home_country string null home_city string null home_airport string null home_airport_lat decimal(38,18) null home_airport_lon decimal(38,18) null sign_up_date date null

In [ ]:
%sql
DESCRIBE sessions;

col_name data_type comment session_id string null user_id int null trip_id string null session_start timestamp null session_end timestamp null flight_discount boolean null hotel_discount boolean null flight_discount_amount decimal(38,18) null hotel_discount_amount decimal(38,18) null flight_booked boolean null hotel_booked boolean null page_clicks int null cancellation boolean null

In [ ]:
%sql
DESCRIBE flights;

col_name data_type comment trip_id string null origin_airport string null destination string null destination_airport string null seats int null return_flight_booked boolean null departure_time timestamp null return_time timestamp null checked_bags int null trip_airline string null destination_airport_lat decimal(38,18) null destination_airport_lon decimal(38,18) null base_fare_usd decimal(38,18) null

In [ ]:
%sql
DESCRIBE hotels;

col_name data_type comment trip_id string null hotel_name string null nights int null rooms int null check_in_time timestamp null check_out_time timestamp null hotel_per_room_usd decimal(38,18) null

**Note**: Column data types were inspected using `DESCRIBE` statements.  
Timestamps, booleans, and numeric fields are consistent with the expected semantics of sessions, users, and products.  
This confirms that the dataset is structurally suitable for reliable joins and lifecycle analysis.

#### 4.2 Primary Keys and Duplicates

In [ ]:
%sql
-- Check duplicated on primary key for sessions
SELECT session_id, COUNT(*) AS duplicates
FROM sessions
GROUP BY session_id
HAVING COUNT(*) > 1;

session_id duplicates

In [ ]:
%sql
-- Check duplicated on primary key for users
SELECT user_id, COUNT(*) AS duplicates
FROM users
GROUP BY user_id
HAVING COUNT(*) > 1;

user_id duplicates

In [ ]:
%sql
-- Check duplicated on primary key for flights
SELECT trip_id, COUNT(*) AS duplicates
FROM flights
GROUP BY trip_id
HAVING COUNT(*) > 1;

trip_id duplicates

In [ ]:
%sql
-- Check duplicated on primary key for hotels
SELECT trip_id, COUNT(*) AS duplicates
FROM hotels
GROUP BY trip_id
HAVING COUNT(*) > 1;

trip_id duplicates

**Note**: Primary keys were validated for uniqueness across all tables:  
`users.user_id`, `sessions.session_id`, `flights.trip_id`, and `hotels.trip_id`.  

No duplicates were detected, ensuring that joins do not introduce artificial row duplication.

#### 4.3 Null Values and Basic Statistics

###### Sessions

In [ ]:
%sql
-- Check null values on sessions columns
SELECT
    COUNT(CASE WHEN session_id IS NULL THEN 1 END) AS session_id_nulls,
    COUNT(CASE WHEN trip_id IS NULL THEN 1 END) AS trip_id_nulls,
    COUNT(CASE WHEN session_start IS NULL THEN 1 END) AS session_start_nulls,
    COUNT(CASE WHEN session_end IS NULL THEN 1 END) AS session_end_nulls,
    COUNT(CASE WHEN page_clicks IS NULL THEN 1 END) AS page_clicks_nulls,
    COUNT(CASE WHEN flight_discount IS NULL THEN 1 END) AS flight_discount_nulls,
    COUNT(CASE WHEN flight_discount_amount IS NULL THEN 1 END) AS flight_discount_amount_nulls,
    COUNT(CASE WHEN flight_booked IS NULL THEN 1 END) AS flight_booked_nulls,
    COUNT(CASE WHEN hotel_discount IS NULL THEN 1 END) AS hotel_discount_nulls,
    COUNT(CASE WHEN hotel_discount_amount IS NULL THEN 1 END) AS hotel_discount_amount_nulls,
    COUNT(CASE WHEN hotel_booked IS NULL THEN 1 END) AS hotel_booked_nulls,
    COUNT(CASE WHEN cancellation IS NULL THEN 1 END) AS cancellation_nulls
FROM sessions;


session_id_nulls trip_id_nulls session_start_nulls session_end_nulls page_clicks_nulls flight_discount_nulls flight_discount_amount_nulls flight_booked_nulls hotel_discount_nulls hotel_discount_amount_nulls hotel_booked_nulls cancellation_nulls 0 3072218 0 0 0 0 4522267 0 0 4716683 0 0

**Note**: Missing values are concentrated in `trip_id` and booking-related fields.  
These NULLs are structurally expected and correspond to browsing-only sessions without an associated trip.  
No unexpected data loss is observed in session records.

###### Users

In [ ]:
%sql
-- Check null values on users columns
SELECT
    COUNT(CASE WHEN user_id IS NULL THEN 1 END) AS user_id_nulls,
    COUNT(CASE WHEN birthdate IS NULL THEN 1 END) AS birthdate_nulls,
    COUNT(CASE WHEN gender IS NULL THEN 1 END) AS gender_nulls,
    COUNT(CASE WHEN married IS NULL THEN 1 END) AS married_nulls,
    COUNT(CASE WHEN has_children IS NULL THEN 1 END) AS has_children_nulls,
    COUNT(CASE WHEN home_country IS NULL THEN 1 END) AS home_country_nulls,
    COUNT(CASE WHEN home_city IS NULL THEN 1 END) AS home_city_nulls,
    COUNT(CASE WHEN home_airport IS NULL THEN 1 END) AS home_airport_nulls,
    COUNT(CASE WHEN home_airport_lat IS NULL THEN 1 END) AS home_airport_lat_nulls,
    COUNT(CASE WHEN home_airport_lon IS NULL THEN 1 END) AS home_airport_lon_nulls,
    COUNT(CASE WHEN sign_up_date IS NULL THEN 1 END) AS sign_up_date_nulls
FROM users;


user_id_nulls birthdate_nulls gender_nulls married_nulls has_children_nulls home_country_nulls home_city_nulls home_airport_nulls home_airport_lat_nulls home_airport_lon_nulls sign_up_date_nulls 0 0 0 0 0 0 0 0 0 0 0

**Note**: The users table has no NULL values in the columns.

###### Flights

In [ ]:
%sql
-- Check null values on flights columns
SELECT
    COUNT(CASE WHEN trip_id IS NULL THEN 1 END) AS trip_id_nulls,
    COUNT(CASE WHEN origin_airport IS NULL THEN 1 END) AS origin_airport_nulls,
    COUNT(CASE WHEN destination IS NULL THEN 1 END) AS destination_nulls,
    COUNT(CASE WHEN destination_airport IS NULL THEN 1 END) AS destination_airport_nulls,
    COUNT(CASE WHEN seats IS NULL THEN 1 END) AS seats_nulls,
    COUNT(CASE WHEN return_flight_booked IS NULL THEN 1 END) AS return_flight_booked_nulls,
    COUNT(CASE WHEN departure_time IS NULL THEN 1 END) AS departure_time_nulls,
    COUNT(CASE WHEN return_time IS NULL THEN 1 END) AS return_time_nulls,
    COUNT(CASE WHEN checked_bags IS NULL THEN 1 END) AS checked_bags_nulls,
    COUNT(CASE WHEN trip_airline IS NULL THEN 1 END) AS trip_airline_nulls,
    COUNT(CASE WHEN destination_airport_lat IS NULL THEN 1 END) AS destination_airport_lat_nulls,
    COUNT(CASE WHEN destination_airport_lon IS NULL THEN 1 END) AS destination_airport_lon_nulls,
    COUNT(CASE WHEN base_fare_usd IS NULL THEN 1 END) AS base_fare_usd_nulls
FROM flights;


trip_id_nulls origin_airport_nulls destination_nulls destination_airport_nulls seats_nulls return_flight_booked_nulls departure_time_nulls return_time_nulls checked_bags_nulls trip_airline_nulls destination_airport_lat_nulls destination_airport_lon_nulls base_fare_usd_nulls 0 0 0 0 0 0 0 88734 0 0 0 0 0

**Note**: `return_time` is NULL for one-way flights. All other flight fields are populated, indicating complete booking information for flights.

###### Hotels

In [ ]:
%sql
-- Check null values on nights columns
SELECT
    COUNT(CASE WHEN trip_id IS NULL THEN 1 END) AS trip_id_nulls,
    COUNT(CASE WHEN hotel_name IS NULL THEN 1 END) AS hotel_name_nulls,
    COUNT(CASE WHEN nights IS NULL THEN 1 END) AS nights_nulls,
    COUNT(CASE WHEN rooms IS NULL THEN 1 END) AS rooms_nulls,
    COUNT(CASE WHEN check_in_time IS NULL THEN 1 END) AS check_in_time_nulls,
    COUNT(CASE WHEN check_out_time IS NULL THEN 1 END) AS check_out_time_nulls,
    COUNT(CASE WHEN hotel_per_room_usd IS NULL THEN 1 END) AS hotel_per_room_usd_nulls
FROM hotels;

trip_id_nulls hotel_name_nulls nights_nulls rooms_nulls check_in_time_nulls check_out_time_nulls hotel_per_room_usd_nulls 0 0 0 0 0 0 0

**Note**: The hotels table has no NULL values in the columns.

### 5. Exploratory Analysis

This section provides a light structured exploratory analysis of the TravelTide tables, with a focus on understanding user behavior, session patterns, and early signals of trip lifecycle dynamics.
The goal is to surface insights that justify later modeling and cohort choices.

#### 5.1 Sessions Table Exploration

##### Total Sessions per User

In [ ]:
%sql
-- Total sessions per user
SELECT user_id, COUNT(*) AS tot_sessions
FROM sessions
GROUP BY user_id
ORDER BY tot_sessions DESC
LIMIT 20;


user_id tot_sessions 36751 17 5655 16 24936 16 6914 16 58289 16 63888 16 29473 16 3389 16 22587 16 49389 16 3745 16 79965 16 7535 16 3814 16 36448 16 15465 15 53425 15 5315 15 36104 15 48020 15

**Note**: The upper tail of the session count distribution was inspected to identify potential extreme user behavior.  
No evident anomalies appear among the most active users, supporting the use of session volume as a valid engagement signal.

##### Cancellation Distribution

In [ ]:
%sql
-- Cancellation distribution in sessions
SELECT cancellation, COUNT(*) AS num_sessions
FROM sessions
GROUP BY cancellation;


cancellation num_sessions true 90670 false 5317393

**Note**: Cancellation sessions represent a small fraction of total sessions.  
This indicates that cancellations are sparse events and should be handled explicitly when reconstructing trip lifecycles.

##### Distinguishing Booking, Browsing, and Cancellation Sessions

In [ ]:
%sql
-- Booking, browsing, and cancellation Sessions
-- Classify sessions based on booking intent and outcome.
SELECT
sum(CASE WHEN (cancellation = false) and (flight_booked=true or hotel_booked=true) THEN 1 ELSE 0 END) as booking_session,
sum(CASE WHEN (cancellation = false) and flight_booked=false and hotel_booked=false THEN 1 ELSE 0 END) as browsing_sessions,
sum(CASE WHEN (cancellation = true) THEN 1 ELSE 0 END) as cancellation_sessions
FROM sessions


booking_session browsing_sessions cancellation_sessions 2245175 3072218 90670

**Note**: The number of browsing_sessions (3,072,218) exactly matches the number of sessions where `trip_id IS NULL`.
This confirms that pure browsing sessions do not have an associated trip, while booking and cancellation sessions do (`trip_id IS NOT NULL`).

This distinction allows us to separate sessions with trips from browsing-only sessions.

##### Trip vs Browsing-Only Sessions

In [ ]:
%sql
-- Trip vs Browsing-Only Sessions
-- Validate the relationship between session intent and presence of trip_id.
SELECT CASE WHEN trip_id IS NULL THEN 'no_trip' ELSE 'with_trip' END AS trip_flag,
       COUNT(*) AS num_sessions
FROM sessions
GROUP BY trip_flag;

trip_flag num_sessions no_trip 3072218 with_trip 2335845

**Note**: This confirms the previous observation: all sessions with a trip have `trip_id` populated, while browsing-only sessions do not.

At this stage, three types of sessions are clearly identified: **browsing-only**, **booking**, and **cancellation**.

These distinctions set the stage for examining whether cancellations share the same `trip_id` as previous booking sessions, for example, to reconstruct the lifecycle of a trip.

##### Booking–Cancellation Relationship

In [ ]:
%sql
-- Test whether cancellation sessions are linked to prior booking sessions via trip_id.
-- Identify trips that appear in cancellation sessions
WITH cancelled_trips AS (
    SELECT DISTINCT trip_id
    FROM sessions
    WHERE cancellation = true
      AND trip_id IS NOT NULL
)

-- Count how many booking sessions share the same trip_id
SELECT
    COUNT(DISTINCT s.trip_id) AS booking_and_cancelled_trips,
    SUM(CASE WHEN s.cancellation = false THEN 1 ELSE 0 END) AS booking_sessions_count,
    SUM(CASE WHEN s.cancellation = true THEN 1 ELSE 0 END) AS cancellation_sessions_count
FROM sessions s
JOIN cancelled_trips ct
  ON s.trip_id = ct.trip_id;

booking_and_cancelled_trips booking_sessions_count cancellation_sessions_count 90670 90670 90670

**Note**: Cancelled trips exhibit two sessions linked by the same trip_id:
**one booking session and one cancellation session**.

This confirms a stable and interpretable trip lifecycle, which is later used to classify sessions
and compute booking and cancellation metrics correctly.

Trips follow a simple and consistent lifecycle:
- **browsing sessions** have no `trip_id`,
- **booking sessions** create a trip (`trip_id`, `cancellation = FALSE`),
- **cancellation sessions** refer back to the same `trip_id` (`cancellation = TRUE`).

This structure clearly distinguishes browsing, booking, and cancellation behavior
and enables coherent session-level extraction without inflating booking counts.

##### Flight/Hotel booking flags in Booking Sessions

In [ ]:
%sql
-- Analyze flight and hotel booking flags in active sessions
SELECT
  flight_booked,
  hotel_booked,
  COUNT(*) AS num_sessions
FROM sessions
WHERE cancellation = false
GROUP BY flight_booked, hotel_booked
ORDER BY flight_booked DESC, hotel_booked DESC;


flight_booked hotel_booked num_sessions true true 1574480 true false 326558 false true 344137 false false 3072218

**Note**: This query counts only non-cancellation sessions (cancellation = false) to capture the real booking activity.

- `flight_booked = true` AND `hotel_booked = true` are sessions where both flight and hotel were booked (1,574,480).

- `flight_booked = true` AND `hotel_booked = false` are flight-only bookings (326,558).

- `flight_booked = false` AND `hotel_booked = true` are hotel-only bookings (344,137).

- `flight_booked = false` AND `hotel_booked = false` are browsing-only sessions with no booking (3,072,218).

**Observation**:
The count of browsing-only sessions (`false, false`) exactly matches the number of sessions with `trip_id = NULL` identified previously, confirming that sessions without a trip correspond to pure browsing activity.

In [ ]:
%sql
-- Check for cancellation sessions with incomplete bookings
SELECT *
FROM sessions
WHERE cancellation = true
  AND (hotel_booked = false OR flight_booked = false);

session_id user_id trip_id session_start session_end flight_discount hotel_discount flight_discount_amount hotel_discount_amount flight_booked hotel_booked page_clicks cancellation

**Note**: This query checks whether there are any cancellation sessions where either `flight_booked` or `hotel_booked` is `false`.

The result, no rows are returned, confirms that all cancellation sessions always have both `flight_booked` and `hotel_booked` set to `true`, regardless of whether the original booking included only a flight, only a hotel, or both.

From a session-level perspective, cancellation flags are standardized; the individual flags no longer reflect the original booking composition. Therefore, when analyzing booking behavior, we must filter out `cancellation = true` to understand the actual booking patterns.

##### Session Details for Trip Shared across Multiple Sessions

In [ ]:
%sql
-- Get all session details for trips shared across multiple sessions
--Identify trips with more than one session
WITH trips_multiple_sessions AS (
    SELECT trip_id
    FROM sessions
    WHERE trip_id IS NOT NULL
    GROUP BY trip_id
    HAVING COUNT(*) > 1
)
-- Retrieve all session records for these trips
SELECT s.*
FROM sessions s
JOIN trips_multiple_sessions tms
    ON s.trip_id = tms.trip_id
ORDER BY s.trip_id, s.session_start
LIMIT 10;

session_id user_id trip_id session_start session_end flight_discount hotel_discount flight_discount_amount hotel_discount_amount flight_booked hotel_booked page_clicks cancellation 1-5da1c1879f5440038967d1405e414ffd 1 1-f35a49334eaa473aa993bde2c8f670ec 2022-02-26T14:19:00.000Z 2022-02-26T14:21:10.000Z false false null null true true 17 false 1-18ef3048214646d8a0d46748b4136401 1 1-f35a49334eaa473aa993bde2c8f670ec 2022-05-31T14:21:10.000Z 2022-05-31T14:46:44.356Z true true null null true true 25 true 100-9a9cb6d0c6584431bd1b59e517e1d60e 100 100-4d42c16dea8441d391a44b1bdddf238b 2021-04-05T16:03:00.000Z 2021-04-05T16:05:28.000Z false false null null true true 20 false 100-e35f4049603f450fa4c83391edcf6580 100 100-4d42c16dea8441d391a44b1bdddf238b 2021-07-16T16:05:28.000Z 2021-07-16T18:05:28.000Z true true null null true true 128 true 1000-de3ba74e7d234d908e36873f69348bd5 1000 1000-824d298be9454790bfa01be81c6bef8d 2021-06-17T15:12:00.000Z 2021-06-17T15:25:50.000Z false false null null true true 111 false 1000-843ed82dc9464c35891ed43819e9eed6 1000 1000-824d298be9454790bfa01be81c6bef8d 2021-06-17T15:25:50.000Z 2021-06-17T17:25:50.000Z true true null null true true 200 true 10001-88a731e9408c49cea4339158a3e59a42 10001 10001-2a94b4fa9e834e9f8703ba8ac65b5bec 2021-06-22T22:53:00.000Z 2021-06-22T22:54:27.000Z true false 0.150000000000000000 null true false 12 false 10001-477c288b465046c49fca4db38c89d517 10001 10001-2a94b4fa9e834e9f8703ba8ac65b5bec 2021-10-31T22:54:27.000Z 2021-11-01T00:54:27.000Z true true null null true true 200 true 100012-211fb2ac2a6941fc8e8e61720d50b2a1 100012 100012-b58d1cc44588426189f46053b68cd4ca 2022-03-07T08:00:00.000Z 2022-03-07T08:02:00.000Z false false null null true false 16 false 100012-6fbecf5f9b484fdd9a715b0a3259d2b5 100012 100012-b58d1cc44588426189f46053b68cd4ca 2022-06-08T08:02:00.000Z 2022-06-08T10:02:00.000Z true true null null true true 197 true

**Note**: Each cancelled trip has exactly two sessions: a booking session and a cancellation session. Cancellation sessions always have all booking flags set to TRUE; to see the actual booking details, refer to the original booking session for the same trip_id.

The behavior of cancellation sessions, in terms of the boolean flags (`flight_booked`, `hotel_booked`, `flight_discount`, `hotel_discount`), will be examined in detail in the session-level notebook.

This deeper analysis will clarify which flags are actually meaningful during cancellations and how to filter the data to capture user behavior in these sessions accurately.

##### Session Duration Distribution

In [ ]:
%sql
-- Distribution of session length
-- Understand how long users typically spend per session.
SELECT
    CASE
        WHEN session_length_minutes BETWEEN 0 AND 5 THEN '0-5 min'
        WHEN session_length_minutes BETWEEN 6 AND 10 THEN '6-10 min'
        WHEN session_length_minutes BETWEEN 11 AND 30 THEN '11-30 min'
        WHEN session_length_minutes BETWEEN 31 AND 60 THEN '31-60 min'
        ELSE '>60 min'
    END AS session_length_bucket,
    COUNT(*) AS num_sessions
FROM (
    SELECT
        CAST((CAST(session_end AS TIMESTAMP) - CAST(session_start AS TIMESTAMP)) AS BIGINT) / 60 AS session_length_minutes
    FROM sessions
) AS sub
GROUP BY session_length_bucket
ORDER BY
    CASE
        WHEN session_length_bucket = '0-5 min' THEN 1
        WHEN session_length_bucket = '6-10 min' THEN 2
        WHEN session_length_bucket = '11-30 min' THEN 3
        WHEN session_length_bucket = '31-60 min' THEN 4
        ELSE 5
    END;


session_length_bucket num_sessions 0-5 min 5053673 6-10 min 128281 11-30 min 28934 31-60 min 11487 >60 min 185688

**Note**: Most sessions (5.05M) are very short, lasting between 0-5 minutes, reflecting typical browsing-only activity. Longer sessions are progressively less frequent: 6-10 minutes (128k), 11-30 minutes (29k), and 31-60 minutes (11k).

Interestingly, a noticeable portion of sessions (>60 minutes, 186k) likely corresponds to more involved activity such as booking trips or researching multiple options.

This distribution highlights the diversity in user engagement and supports the distinction between quick browsing sessions and full booking sessions in further analysis.

In [ ]:
%sql
-- Count page clicks in session
SELECT page_clicks, COUNT(*) AS count
FROM sessions
GROUP BY page_clicks
ORDER BY page_clicks;

page_clicks count 0 7 1 40921 2 183693 3 260132 4 274756 5 260093 6 234807 7 207309 8 186882 9 186806 10 192506 11 180060 12 163467 13 145805 14 128668 15 119635 16 141716 17 185663 18 209234 19 206226 20 191308 21 170927 22 152212 23 133566 24 118179 25 103646 26 91012 27 80794 28 71769 29 63915 30 56586 31 50513 32 44899 33 40389 34 36548 35 33212 36 29965 37 26755 38 24586 39 22446 40 20491 41 18622 42 17310 43 15810 44 14720 45 13280 46 12297 47 11426 48 10367 49 9640 50 8948 51 8403 52 7851 53 7312 54 6729 55 6240 56 5976 57 5491 58 5104 59 4777 60 4633 61 4291 62 3999 63 3684 64 3496 65 3282 66 3146 67 2995 68 2905 69 2711 70 2552 71 2400 72 2380 73 2160 74 2120 75 1962 76 1894 77 1799 78 1714 79 1691 80 1583 81 1518 82 1454 83 1340 84 1216 85 1302 86 1232 87 1103 88 1154 89 1078 90 1026 91 941 92 979 93 937 94 920 95 860 96 841 97 811 98 736 99 759 100 705 101 727 102 658 103 681 104 590 105 581 106 622 107 576 108 541 109 537 110 555 111 580 112 505 113 494 114 479 115 477 116 443 117 463 118 431 119 396 120 438 121 387 122 407 123 384 124 389 125 346 126 342 127 335 128 358 129 355 130 320 131 313 132 336 133 308 134 306 135 298 136 298 137 323 138 296 139 240 140 242 141 248 142 245 143 243 144 243 145 251 146 229 147 233 148 271 149 232 150 221 151 220 152 236 153 231 154 210 155 197 156 233 157 193 158 200 159 220 160 199 161 187 162 211 163 202 164 177 165 190 166 199 167 174 168 204 169 199 170 187 171 176 172 145 173 183 174 159 175 175 176 162 177 156 178 167 179 157 180 158 181 172 182 151 183 181 184 153 185 139 186 153 187 134 188 141 189 140 190 148 191 155 192 144 193 171 194 142 195 135 196 118 197 137 198 123 199 130 200 38345 201 16 202 17 203 16 204 14 205 6 206 15 207 16 208 10 209 10 210 12 211 10 212 9 213 9 214 8 215 13 216 8 217 15 218 10 219 12 220 8 221 15 222 18 223 9 224 5 225 6 226 5 227 9 228 9 229 12 230 7 231 15 232 10 233 15 234 8 235 13 236 6 237 8 238 9 239 8 240 8 241 11 242 8 243 10 244 7 245 8 246 2 247 3 248 9 249 10 250 5 251 5 252 8 253 3 254 2 255 4 256 3 257 3 258 6 259 4 260 1 261 1 262 5 263 9 264 1 265 3 266 3 267 1 268 2 269 3 270 2 271 2 272 4 273 4 274 2 275 3 276 3 277 2 278 2 279 5 280 1 281 2 282 4 284 3 285 2 286 3 287 5 289 6 292 3 293 1 294 1 295 1 296 3 298 5 299 1 300 1 301 2 302 4 305 2 306 2 307 4 309 2 311 1 312 3 313 1 314 1 315 1 316 1 317 1 319 1 322 2 324 3 325 3 326 1 327 1 328 2 329 1 330 1 331 2 332 1 333 1 336 1 337 2 338 3 340 1 343 1 344 1 346 1 349 3 351 2 352 1 353 1 354 1 355 2 360 1 361 1 362 1 364 1 368 1 370 2 373 1 374 1 383 1 387 1 392 1 395 1 396 1 397 1 399 1 400 2 402 1 407 2 408 1 412 1 418 1 419 1 426 1 440 1 448 1 451 1 455 1 461 1 465 1 479 1 484 1 502 1 515 1 533 1 537 1 566 1 573 1 630 1 755 1 757 1 845 1 872 1 874 1 893 1 945 1 1555 1 1912 1 2421 1

**Note**: The distribution highlights the presence of a large number of browsing-only sessions,
which explains the high frequency of NULL values in `trip_id` and booking-related fields observed earlier.

##### Discount Offered Flags

In [ ]:
%sql
SELECT
    flight_discount,
    hotel_discount,
    COUNT(*) AS num_sessions
FROM sessions
WHERE cancellation = false
GROUP BY flight_discount, hotel_discount
ORDER BY flight_discount DESC, hotel_discount DESC;

flight_discount hotel_discount num_sessions true true 137032 true false 748764 false true 554348 false false 3877249

**Note**: This table shows the distribution of sessions according to whether flight and hotel discounts were offered. Only non-cancellation sessions are considered to avoid distortion caused by cancellation sessions, where all discount flags are automatically set to TRUE.

- Most sessions (3.87M) did not have any discounts applied.

- Flight-only discounts appear in ~748k sessions, hotel-only discounts in 554k sessions.

- Both flight and hotel discounts were offered in 137k sessions.

This gives us a realistic picture of promotional exposure across active user sessions, which will be relevant for understanding user behavior and price sensitivity in later analyses.

In [ ]:
%sql
SELECT
    u.gender,

    COUNT(*) AS total_sessions,

    SUM(
        CASE
            WHEN s.cancellation = FALSE
             AND (s.flight_booked = TRUE OR s.hotel_booked = TRUE)
            THEN 1 ELSE 0
        END
    ) AS booking_sessions,

    SUM(
        CASE
            WHEN s.cancellation = TRUE
            THEN 1 ELSE 0
        END
    ) AS cancellation_sessions,

    -- Conversion rate: bookings / total sessions
    ROUND(
        SUM(
            CASE
                WHEN s.cancellation = FALSE
                 AND (s.flight_booked = TRUE OR s.hotel_booked = TRUE)
                THEN 1 ELSE 0
            END
        ) * 1.0 / COUNT(*),
        3
    ) AS booking_rate,

    -- Cancellation rate: cancellations / booking sessions
    ROUND(
        SUM(
            CASE
                WHEN s.cancellation = TRUE
                THEN 1 ELSE 0
            END
        ) * 1.0 /
        NULLIF(
            SUM(
                CASE
                    WHEN s.cancellation = FALSE
                     AND (s.flight_booked = TRUE OR s.hotel_booked = TRUE)
                    THEN 1 ELSE 0
                END
            ),
            0
        ),
        3
    ) AS cancellation_rate

FROM sessions s
JOIN users u
  ON s.user_id = u.user_id

GROUP BY u.gender
ORDER BY total_sessions DESC;


gender total_sessions booking_sessions cancellation_sessions booking_rate cancellation_rate M 2722001 1162748 44789 0.427 0.039 F 2645784 1065060 45227 0.403 0.042 O 40278 17367 654 0.431 0.038

In [ ]:
%sql
SELECT
    gender,
    MIN(session_start) AS first_session_date,
    MAX(session_start) AS last_session_date,
    COUNT(DISTINCT s.user_id) AS n_users
FROM sessions s
JOIN users u ON s.user_id = u.user_id
GROUP BY gender
ORDER BY gender;


gender first_session_date last_session_date n_users F 2021-04-01T04:57:00.000Z 2023-07-28T23:24:52.000Z 453654 M 2021-04-01T00:42:00.000Z 2023-07-28T23:57:55.000Z 558986 O 2021-04-11T14:54:00.000Z 2023-07-28T18:17:47.000Z 8286

#### 5.2 Users Table Exploration

##### Distribution by Gender, Marital Status, and Children

In [ ]:
%sql
-- Distribution by gender, marital status, and children
SELECT gender, married, has_children, COUNT(user_id) AS tot_users
FROM users
GROUP BY gender, married, has_children;

gender married has_children tot_users M true true 82754 M false true 91457 F true false 110504 M true false 127701 O false true 1659 O true true 670 F false true 70928 M false false 257074 F true true 72169 O true false 1019 F false false 200053 O false false 4938

**Note**: The user base is diverse across gender, marital status, and parental status.

- Male users (M) represent a slightly larger segment than female users (F), while O (Other/unspecified) is very small.

- Most users without children are single or married (M false/true), while users with children are more balanced across gender and marital status.

- This distribution highlights that family-related features (has_children and married) could be meaningful in customer segmentation.

- No extreme outliers are apparent in user counts per category, suggesting a reasonably balanced demographic representation.

##### Age at Travel Distribution

In [ ]:
%sql
-- Age distribution retrieved from users based on date of Travel
SELECT
  YEAR(u.birthdate) AS birthyear,
  FLOOR(datediff(f.departure_time, u.birthdate) / 365) AS age_at_travel,
  COUNT(DISTINCT u.user_id) AS tot_users
FROM users u
JOIN sessions s ON s.user_id = u.user_id
JOIN flights f ON f.trip_id = s.trip_id
GROUP BY
  YEAR(u.birthdate),
  FLOOR(datediff(f.departure_time, u.birthdate) / 365)
ORDER BY birthyear DESC;

birthyear age_at_travel tot_users 2006 14 194 2006 15 3359 2006 16 14421 2006 18 636 2006 17 10249 2005 19 114 2005 15 38 2005 16 608 2005 17 2431 2005 18 1719 2004 17 690 2004 20 146 2004 18 2815 2004 19 2007 2004 16 40 2003 20 2232 2003 21 163 2003 19 3130 2003 17 38 2003 18 733 2002 20 3626 2002 21 2609 2002 22 177 2002 19 885 2002 18 61 2001 21 3957 2001 19 45 2001 20 960 2001 22 2768 2001 23 202 2000 21 1001 2000 22 4265 2000 23 3011 2000 20 60 2000 24 210 1999 22 1128 1999 23 4808 1999 25 227 1999 21 70 1999 24 3531 1998 24 5250 1998 23 1210 1998 25 3849 1998 26 252 1998 22 70 1997 26 4895 1997 23 175 1997 25 9114 1997 27 182 1997 24 2388 1996 24 265 1996 26 12132 1996 25 3290 1996 27 5643 1996 28 153 1995 25 287 1995 27 12670 1995 26 3454 1995 29 159 1995 28 5996 1994 28 13746 1994 29 6442 1994 30 166 1994 26 259 1994 27 3673 1993 28 3917 1993 27 335 1993 31 197 1993 29 14245 1993 30 6853 1992 28 336 1992 31 7367 1992 30 15347 1992 29 4064 1992 32 192 1991 30 4185 1991 31 15809 1991 32 7609 1991 33 226 1991 29 328 1990 33 8083 1990 34 220 1990 32 16730 1990 31 4529 1990 30 337 1989 35 243 1989 33 17376 1989 31 350 1989 32 4525 1989 34 8258 1988 34 18253 1988 36 223 1988 35 8535 1988 33 4855 1988 32 356 1987 37 186 1987 34 5444 1987 36 9374 1987 33 450 1987 35 20244 1986 34 491 1986 38 170 1986 35 5915 1986 36 21414 1986 37 9593 1985 35 460 1985 38 9988 1985 36 5861 1985 37 22075 1985 39 145 1984 40 174 1984 37 6127 1984 36 496 1984 39 10282 1984 38 22407 1983 41 166 1983 39 22092 1983 37 531 1983 38 6048 1983 40 10225 1982 42 175 1982 40 21980 1982 41 10041 1982 38 470 1982 39 5961 1981 39 463 1981 43 167 1981 40 5852 1981 42 10024 1981 41 21753 1980 44 162 1980 43 9822 1980 40 468 1980 41 5858 1980 42 21727 1979 42 5723 1979 45 166 1979 41 444 1979 44 9739 1979 43 21100 1978 44 20171 1978 43 5456 1978 45 9473 1978 42 476 1978 46 171 1977 46 9246 1977 44 5199 1977 43 415 1977 45 19730 1977 47 184 1976 48 154 1976 47 8758 1976 45 5111 1976 46 18810 1976 44 438 1975 48 8348 1975 45 402 1975 47 18055 1975 46 4852 1975 49 138 1974 49 8021 1974 48 16921 1974 47 4514 1974 46 365 1974 50 139 1973 50 7663 1973 49 16175 1973 48 4343 1973 51 160 1973 47 344 1972 49 3939 1972 48 314 1972 50 15130 1972 52 127 1972 51 7200 1971 50 3771 1971 52 6555 1971 49 297 1971 53 117 1971 51 14166 1970 52 13159 1970 54 128 1970 53 6239 1970 50 257 1970 51 3518 1969 54 5680 1969 53 12156 1969 52 3230 1969 55 96 1969 51 270 1968 55 5439 1968 53 2972 1968 52 214 1968 54 11178 1968 56 102 1967 54 2495 1967 57 143 1967 53 216 1967 56 4587 1967 55 9287 1966 55 1826 1966 56 7240 1966 54 163 1966 57 3884 1966 58 148 1965 57 5957 1965 56 1533 1965 55 91 1965 58 3233 1965 59 138 1964 57 1428 1964 59 2909 1964 56 94 1964 60 111 1964 58 5529 1963 60 1100 1963 57 33 1963 59 2040 1963 61 39 1963 58 528 1962 59 53 1962 61 119 1962 60 194 1962 62 6 1962 58 3 1961 59 27 1961 61 1374 1961 63 25 1961 62 765 1961 60 349 1960 62 2531 1960 61 613 1960 64 51 1960 63 1422 1960 60 39 1959 65 83 1959 63 3122 1959 61 58 1959 64 1697 1959 62 777 1958 66 68 1958 65 1812 1958 63 850 1958 64 3233 1958 62 57 1957 65 2300 1957 67 94 1957 66 1500 1957 63 31 1957 64 506 1956 68 97 1956 65 269 1956 66 1293 1956 67 1231 1956 64 9 1955 65 10 1955 69 87 1955 66 224 1955 67 1124 1955 68 1084 1954 66 13 1954 70 81 1954 68 1011 1954 69 888 1954 67 208 1953 71 75 1953 70 832 1953 68 173 1953 67 10 1953 69 892 1952 72 60 1952 71 668 1952 69 161 1952 70 802 1952 68 9 1951 73 42 1951 70 114 1951 72 652 1951 71 601 1951 69 6 1950 73 501 1950 74 39 1950 71 113 1950 72 541 1950 70 5 1949 72 76 1949 71 3 1949 74 431 1949 73 441 1949 75 40 1948 74 346 1948 76 31 1948 75 344 1948 73 63 1948 72 4 1947 74 61 1947 76 289 1947 75 289 1947 77 31 1947 73 1 1946 78 18 1946 76 251 1946 77 251 1946 75 49 1946 74 2 1945 77 157 1945 79 12 1945 76 46 1945 78 198 1945 75 2 1944 78 164 1944 79 166 1944 80 15 1944 77 33 1943 81 9 1943 79 133 1943 78 19 1943 80 101 1943 77 2 1942 79 16 1942 80 104 1942 82 6 194

**Note**: The age-at-travel distribution, useful for cohort analysis,  shows a strong concentration in the adult population, with the highest number of users clustered between approximately 25 and 55 years old. Younger users (below 18) and older age groups (above 70) are present but represent a significantly smaller share of the total population.

Overall, the distribution appears smooth and consistent across birth cohorts, suggesting no major anomalies in the age calculation and a user base primarily composed of working-age travelers.

In [ ]:
%sql
SELECT home_country, COUNT(user_id) as tot_users
FROM users
GROUP BY home_country;

home_country tot_users canada 172572 usa 848354

**Note**: The user distribution is strongly skewed toward North America, with most users coming from the United States and a smaller but still significant share from Canada. This geographic  concentration may provide useful context for interpreting user behavior in subsequent analyses.

#### 5.3 Flights and Hotels Exploration

##### Most used airlines in the last 6 months

In [ ]:
%sql
-- Most used airlines in the last 6 months
SELECT trip_airline, COUNT(trip_id) AS tot_used
FROM flights
WHERE departure_time > (SELECT MAX(departure_time) - INTERVAL '6 months' FROM flights)
GROUP BY trip_airline
ORDER BY tot_used DESC
LIMIT 30;

trip_airline tot_used American Airlines 2590 Delta Air Lines 2437 United Airlines 2288 Ryanair 1731 Southwest Airlines 987 JetBlue Airways 785 Allegiant Air 731 Air Canada 727 AirTran Airways 592 Air India Limited 549 Air China 544 China Southern Airlines 517 Alitalia 437 Turkish Airlines 427 WestJet 421 All Nippon Airways 406 easyJet 387 Japan Airlines 384 Air France 383 China Eastern Airlines 375 KLM Royal Dutch Airlines 335 Emirates 314 Cathay Pacific 313 Thai Airways International 308 Lufthansa 302 AirAsia 298 Malaysia Airlines 297 Shenzhen Airlines 267 Alaska Airlines 264 Vietnam Airlines 263

**Note**: The airline usage distribution is highly skewed toward a small set of major carriers, with American Airlines, Delta Air Lines, and United Airlines being the most frequently used.

Low-cost and regional airlines are also well represented, indicating a diverse mix of travel patterns and user preferences across both full-service and budget carriers.

##### Average Seats by Airline

In [ ]:
%sql
-- average seats used by airline
SELECT trip_airline, AVG(seats) AS average_seats
FROM flights
GROUP BY trip_airline
ORDER BY average_seats DESC;

trip_airline average_seats Binter Canarias 3.3333333333333335 LSM Airlines 3.1666666666666665 Helvetic Airways 3.0 Air Greenland 3.0 Hapagfly 2.6666666666666665 Israir 2.5 Kam Air 2.5 Air Bosna 2.5 Arkia Israel Airlines 2.4615384615384617 TRIP Linhas A 2.4545454545454546 Montenegro Airlines 2.3125 Tuninter 2.2857142857142856 Air Niugini 2.282608695652174 Air Florida 2.2666666666666666 Marusya Airways 2.25 Flybaboo 2.2333333333333334 Intersky 2.230769230769231 Rwandair Express 2.2083333333333335 Empresa Ecuatoriana De Aviacion 2.1954022988505746 Hokkaido International Airlines 2.1666666666666665 Varig Log 2.1666666666666665 Corse-Mediterranee 2.1636363636363636 Nouvel Air Tunisie 2.1538461538461537 Air Busan 2.1443298969072164 Atlant-Soyuz Airlines 2.142857142857143 Airlinair 2.1363636363636362 Royal Brunei Airlines 2.135135135135135 Golden Air 2.1323529411764706 Interlink Airlines 2.1320754716981134 Air Moldova 2.125 Daallo Airlines 2.125 Cargo Plus Aviation 2.123076923076923 Aurigny Air Services 2.090909090909091 Blue Panorama Airlines 2.086206896551724 Mandala Airlines 2.0849056603773586 Belavia Belarusian Airlines 2.0786516853932584 Skymark Airlines 2.0720720720720722 Air Astana 2.0714285714285716 Nas Air 2.066225165562914 Croatia Airlines 2.0654205607476634 Homer Air 2.061224489795918 Saratov Aviation Division 2.0555555555555554 Jetstar Asia Airways 2.0526315789473686 Luxair 2.051282051282051 Mandarin Airlines 2.05 Oman Air 2.0428571428571427 Estonian Air 2.037037037037037 Aserca Airlines 2.0368663594470044 Yemenia 2.0358974358974358 Air Arabia 2.0314465408805034 Shaheen Air International 2.0283018867924527 Malmö Aviation 2.024390243902439 Thomsonfly 2.018292682926829 Belair Airlines 2.017543859649123 Pegasus Airlines 2.015625 Air Serbia 2.0142857142857142 Spicejet 2.009693053311793 Air Madagascar 2.0078740157480315 Hong Kong Airlines 2.003129890453834 Go Air 2.002958579881657 Jazeera Airways 2.0 Air Burkina 2.0 Air Botswana 2.0 Widerøe 2.0 South East Asian Airlines 2.0 Aeroflot-Nord 2.0 Air India Express 2.0 Air Sinai 2.0 Dennis Sky 2.0 Atlantic Airways 2.0 Air Zimbabwe 2.0 Santa Barbara Airlines 2.0 Salzburg arrows 2.0 Air Mauritius 1.99 Shenzhen Airlines 1.9883578431372548 Air Sahara 1.9863481228668942 Shandong Airlines 1.9853801169590644 Czech Airlines 1.985096870342772 Air Bourbon 1.9819277108433735 AeroWorld 1.981818181818182 Niki 1.9793814432989691 China SSS 1.9763948497854078 Ibex Airlines 1.9763513513513513 Druk Air 1.9753086419753085 Germanwings 1.974074074074074 Juneyao Airlines 1.9736842105263157 Sudan Airways 1.972972972972973 Air Foyle 1.972027972027972 Air One 1.9714285714285715 Thai AirAsia 1.9702970297029703 TAAG Angola Airlines 1.9696969696969697 Ural Airlines 1.9669421487603307 Middle East Airlines 1.9663865546218486 Air Austral 1.9642857142857142 Mahan Air 1.962962962962963 Maxair 1.9622641509433962 Onur Air 1.9593495934959348 Shanghai Airlines 1.9567147613762486 Dragonair 1.9557675628794449 Meridiana 1.953781512605042 TransAsia Airways 1.952802359882006 VRG Linhas Aereas 1.952662721893491 TAME 1.948051948051948 Sriwijaya Air 1.9470198675496688 Vietnam Airlines 1.9463579898362506 Travel Service 1.9387755102040816 Wizz Air 1.937984496124031 Japan Transocean Air 1.9375 Jetstar Airways 1.9354005167958657 Volotea Costa Rica 1.9345372460496615 Flybe 1.9342105263157894 TUIfly 1.9335038363171355 Air Malta 1.9333333333333333 MIAT Mongolian Airlines 1.9333333333333333 Hex'Air 1.9333333333333333 Garuda Indonesia 1.9328621908127208 L 1.9307692307692308 Kenya Airways 1.9303797468354431 Thai Airways International 1.9302540415704388 Thomas Cook Airlines 1.9291044776119404 IndiGo Airlines 1.9287020109689215 Germania 1.9262295081967213 AirAsia 1.9244984160506864 Ariana Afghan Airlines 1.9230769230769231 easyJet 1.9230205278592376 Turkmenistan Airlines 1.9166666666666667 Rainbow Air Polynesia 1.9166666666666667 Transavia Holland 1.9154013015184381 SilkAir 1.9153094462540716 Potomac Air 1.911764705882353 Eurolot 1.911111

**Note**: Airlines with a lower average number of seats per trip are mostly associated with business travel or bookings for one or two passengers.

In contrast, airlines showing higher average seat counts tend to be preferred for family or group travel, where price sensitivity is higher and more economical carriers are often selected.

##### Hotel Name Distribution

In [ ]:
%sql
-- Hotel name distribution
SELECT hotel_name, COUNT(*) AS tot_hotel_names
FROM hotels
GROUP BY hotel_name
ORDER BY tot_hotel_names DESC;

hotel_name tot_hotel_names Extended Stay - new york 14075 Radisson - new york 14073 Starwood - new york 14029 Conrad - new york 14022 Rosewood - new york 14017 Banyan Tree - new york 13974 Best Western - new york 13959 Shangri-La - new york 13958 InterContinental - new york 13956 Aman Resorts - new york 13940 Hyatt - new york 13940 Accor - new york 13915 Four Seasons - new york 13900 Hilton - new york 13839 Wyndham - new york 13828 NH Hotel - new york 13826 Marriott - new york 13799 Crowne Plaza - new york 13795 Choice Hotels - new york 13745 Fairmont - new york 13744 Four Seasons - los angeles 7185 Banyan Tree - los angeles 7121 Wyndham - los angeles 7073 Extended Stay - los angeles 7060 Conrad - los angeles 7052 Best Western - los angeles 7050 Rosewood - los angeles 7044 Choice Hotels - los angeles 7027 Radisson - los angeles 7021 NH Hotel - los angeles 7017 Fairmont - los angeles 7006 Marriott - los angeles 6979 Aman Resorts - los angeles 6976 Hilton - los angeles 6975 Starwood - los angeles 6969 Crowne Plaza - los angeles 6969 InterContinental - los angeles 6920 Shangri-La - los angeles 6909 Hyatt - los angeles 6867 Accor - los angeles 6816 Starwood - toronto 5218 Rosewood - toronto 5189 Shangri-La - toronto 5157 Banyan Tree - toronto 5138 Marriott - toronto 5111 Radisson - toronto 5098 Crowne Plaza - toronto 5085 Best Western - toronto 5082 Extended Stay - toronto 5077 Accor - toronto 5075 Fairmont - toronto 5061 Hilton - toronto 5054 Aman Resorts - toronto 5041 Hyatt - toronto 5018 Four Seasons - toronto 4996 Conrad - toronto 4991 Hyatt - chicago 4957 Starwood - chicago 4955 NH Hotel - toronto 4942 InterContinental - toronto 4933 Wyndham - toronto 4930 Radisson - chicago 4928 Best Western - chicago 4927 Rosewood - chicago 4920 Fairmont - chicago 4914 Choice Hotels - toronto 4896 Extended Stay - chicago 4896 Shangri-La - chicago 4895 Aman Resorts - chicago 4884 Marriott - chicago 4883 Wyndham - chicago 4879 Hilton - chicago 4869 Crowne Plaza - chicago 4864 Four Seasons - chicago 4840 Accor - chicago 4815 Banyan Tree - chicago 4805 InterContinental - chicago 4796 Conrad - chicago 4794 Choice Hotels - chicago 4751 NH Hotel - chicago 4736 Crowne Plaza - houston 4212 Marriott - houston 4201 Accor - houston 4187 Choice Hotels - houston 4173 Hilton - houston 4170 Starwood - houston 4156 Wyndham - houston 4154 Shangri-La - houston 4151 Extended Stay - houston 4147 Fairmont - houston 4144 Rosewood - houston 4130 Hyatt - houston 4124 Four Seasons - houston 4099 Best Western - houston 4090 Conrad - houston 4086 Aman Resorts - houston 4083 Banyan Tree - houston 4061 Radisson - houston 4050 InterContinental - houston 4049 NH Hotel - houston 4048 Starwood - montreal 3256 NH Hotel - montreal 3236 Extended Stay - montreal 3232 Marriott - montreal 3228 Rosewood - montreal 3213 Four Seasons - montreal 3211 Accor - montreal 3199 Shangri-La - montreal 3193 Wyndham - montreal 3184 Choice Hotels - montreal 3177 Banyan Tree - montreal 3175 Hilton - montreal 3159 Best Western - montreal 3156 Radisson - montreal 3151 Aman Resorts - montreal 3148 Fairmont - montreal 3125 Crowne Plaza - montreal 3122 Conrad - montreal 3107 Hyatt - montreal 3097 InterContinental - montreal 3084 Accor - phoenix 2956 Aman Resorts - philadelphia 2925 Fairmont - phoenix 2923 Radisson - phoenix 2919 Rosewood - philadelphia 2895 InterContinental - philadelphia 2895 Banyan Tree - philadelphia 2892 InterContinental - phoenix 2882 Wyndham - philadelphia 2880 Best Western - philadelphia 2876 Extended Stay - philadelphia 2873 NH Hotel - philadelphia 2868 Aman Resorts - phoenix 2860 Fairmont - philadelphia 2859 Hilton - philadelphia 2858 Radisson - philadelphia 2855 Conrad - phoenix 2852 Choice Hotels - phoenix 2850 Hyatt - philadelphia 2847 NH Hotel - phoenix 2840 Shangri-La - phoenix 2835 Shangri-La - philadelphia 2833 Starwood - phoenix 2826 Conrad - philadelphia 2825 Hilton - phoenix 2822 Choice Hotels - philadelphia 2817 Accor - philadelphia 2816 Best Western - phoenix 28

**Note**: Hotel usage is heavily concentrated in major North American cities, with New York, Los Angeles, Toronto, and Chicago showing the highest number of stays across almost all hotel chains. Large international brands (e.g. Marriott, Hilton, Hyatt, Accor) appear consistently across locations, indicating strong brand standardization and user preference for well-known chains. Smaller volumes are observed in international cities, suggesting more localized or less frequent travel compared to domestic destinations.

In [ ]:
%sql
--Annual Flight Counts Based on Departure Year
SELECT
    EXTRACT(YEAR FROM f.departure_time) AS year,
    COUNT(*) AS num_flights
FROM flights f
WHERE f.departure_time IS NOT NULL
GROUP BY 1
ORDER BY 1;

year num_flights 2021 83346 2022 634052 2023 1145121 2024 38519

**Note**: The distribution of flight departures is strongly concentrated in 2023, which contains the majority of observed records. This reinforces the relevance of focusing the analysis on a recent and well-defined temporal window, consistent with Elena’s reference timeframe. Earlier and later years are present but significantly less represented.

##### Nights Distribution

In [ ]:
%sql
-- Count hotel sessions grouped by hotel nights
SELECT
    nights,
    COUNT(*) AS num_sessions
FROM hotels
GROUP BY nights
ORDER BY nights;

nights num_sessions -2 281 -1 11786 0 139952 1 378046 2 342120 3 250151 4 177756 5 128172 6 95335 7 73353 8 58765 9 48550 10 40212 11 33503 12 27821 13 22537 14 17987 15 14750 16 11558 17 9150 18 7175 19 5513 20 4520 21 3627 22 2831 23 2308 24 1848 25 1480 26 1254 27 959 28 776 29 643 30 580 31 485 32 380 33 351 34 276 35 228 36 209 37 158 38 166 39 135 40 118 41 87 42 94 43 76 44 70 45 62 46 49 47 25 48 38 49 29 50 31 51 27 52 26 53 11 54 19 55 15 56 12 57 17 58 11 59 18 60 7 61 6 62 2 63 2 64 7 65 8 66 5 67 8 68 1 69 4 70 9 71 4 72 2 73 5 74 3 75 2 77 1 78 3 79 2 80 2 82 1 85 1 86 1 91 1 94 1 95 1 98 2 99 1 105 2 107 1

**Note**: The distribution of nights is heavily right-skewed, with the majority of sessions concentrated between 1 and 3 nights. The frequency decreases steadily as the number of nights increases, indicating that short stays are far more common than long ones.

Negative and zero values are present and likely represent data quality issues or edge cases; these will be addressed in the next notebook during data cleaning.

In [ ]:
%sql
-- Annual Hotel Counts Based on Check-in Year
SELECT
    EXTRACT(YEAR FROM h.check_in_time) AS year,
    COUNT(*) AS num_stays
FROM hotels h
WHERE h.check_in_time IS NOT NULL
GROUP BY 1
ORDER BY 1;

year num_stays 2021 88781 2022 661387 2023 1153086 2024 15363

**Note**: The distribution of hotel check-ins mirrors that of flights, with the majority of records concentrated in 2023. This confirms that the main activity falls within a recent and well-defined temporal window, supporting the focus on this period for further analysis.

#### 5.4 Cohort Sensitivity Analysis (User-Level)

This section evaluates the robustness of the user cohort definition used for subsequent analysis.
In particular, it examines how different minimum-session thresholds affect:

- User representation by gender

- The stability of booking and cancellation behavior

The goal is to ensure that the selected cohort provides sufficient behavioral depth **without introducing systematic bias**.

##### 5.4.1 Baseline Cohort

###### Cohort definition: sessions > 7

In [ ]:
%sql
-- select sessions with session_start > '2023-01-04'
WITH filtered_sessions AS (
  SELECT
    user_id,
    trip_id,
    cancellation,
    flight_booked,
    hotel_booked
  FROM sessions
  WHERE session_start > '2023-01-04'
),

-- select eligible users from filtered_sessions
eligible_users AS (
  SELECT user_id
  FROM filtered_sessions
  GROUP BY user_id
  HAVING COUNT(*) > 7  -- users with more than 7 sessions after 2023-01-04
),

-- select cohort users in filtered_sessions
cohort_sessions AS (
  SELECT fs.*
  FROM filtered_sessions fs
  JOIN eligible_users eu ON eu.user_id = fs.user_id
),

-- select all sessions from cohort sessions
sessions_by_user AS (
  SELECT
    user_id,
    COUNT(*) AS total_sessions,
    COUNT(*) FILTER (WHERE trip_id IS NULL) AS browsing_sessions, -- only browsing sessions
    COUNT(*) FILTER (
      WHERE trip_id IS NOT NULL AND cancellation = FALSE
      AND (flight_booked = TRUE OR hotel_booked = TRUE) -- only booking sessions
    ) AS booking_sessions,
    COUNT(*) FILTER (WHERE cancellation = TRUE) AS cancellation_sessions -- only cancellation sessions
  FROM cohort_sessions
  GROUP BY user_id
),

/* select user by gender from filtered sessions_by_user
now to reduce the query execution time */
users_with_gender AS (
  SELECT
    sbu.*,
    u.gender
  FROM sessions_by_user sbu
  JOIN users u ON u.user_id = sbu.user_id
)

-- select all the information from users_with_gender and perform aggregations
SELECT
  gender,
  COUNT(*) AS n_users,
  SUM(browsing_sessions) AS n_browsings, -- total number of browsing sessions
  SUM(booking_sessions) AS n_bookings,   -- total number of booking sessions
  SUM(cancellation_sessions) AS n_cancels, -- total number of cancellations
  SUM(total_sessions) AS n_sessions, -- total number of sessions
  ROUND(SUM(booking_sessions)::numeric / NULLIF(SUM(total_sessions), 0), 3) AS booking_rate, -- booking rate
  ROUND(SUM(cancellation_sessions)::numeric / NULLIF(SUM(booking_sessions), 0), 3) AS cancel_rate -- cancellation rate
FROM users_with_gender
GROUP BY gender -- group by gender
ORDER BY n_users DESC;


gender n_users n_browsings n_bookings n_cancels n_sessions booking_rate cancel_rate F 5292 28741 14224 521 43486 0.327 0.037 M 695 3708 1839 88 5635 0.326 0.048 O 11 60 29 1 90 0.322 0.034

**Note**: This analysis evaluates the sensitivity of user-level metrics to different minimum-session cohort thresholds.

`n_sessions` includes all session types (browsing, booking, cancellation).  
`n_bookings` counts booking sessions only (`trip_id IS NOT NULL` and `cancellation = FALSE`).  
`n_cancels` counts cancellation sessions (`cancellation = TRUE`).


`booking_rate` is computed as `n_bookings / n_sessions`, while `cancellation_rate` is computed as `n_cancels / n_bookings`.

Comparing multiple minimum-activity thresholds shows that absolute gender representation varies with the cohort definition, while booking and cancellation rates remain stable.  
This indicates that the selected cohort (>7 sessions) preserves behavioral validity while prioritizing users with sufficient interaction history.

##### 5.4.2 Lower Threshold Cohort

###### Cohort definition: sessions > 5

In [ ]:
%sql
-- select sessions with session_start > '2023-01-04'
WITH filtered_sessions AS (
  SELECT
    user_id,
    trip_id,
    cancellation,
    flight_booked,
    hotel_booked
  FROM sessions
  WHERE session_start > '2023-01-04'
),

-- select eligible users from filtered_sessions
eligible_users AS (
  SELECT user_id
  FROM filtered_sessions
  GROUP BY user_id
  HAVING COUNT(*) > 5  -- users with more than 5 sessions after 2023-01-04
),

-- select cohort users in filtered_sessions
cohort_sessions AS (
  SELECT fs.*
  FROM filtered_sessions fs
  JOIN eligible_users eu ON eu.user_id = fs.user_id
),

-- select all sessions from cohort sessions
sessions_by_user AS (
  SELECT
    user_id,
    COUNT(*) AS total_sessions,
    COUNT(*) FILTER (WHERE trip_id IS NULL) AS browsing_sessions, -- only browsing sessions
    COUNT(*) FILTER (
      WHERE trip_id IS NOT NULL AND cancellation = FALSE
      AND (flight_booked = TRUE OR hotel_booked = TRUE) -- only booking sessions
    ) AS booking_sessions,
    COUNT(*) FILTER (WHERE cancellation = TRUE) AS cancellation_sessions -- only cancellation sessions
  FROM cohort_sessions
  GROUP BY user_id
),

/* select user by gender from filtered sessions_by_user
now to reduce the query execution time */
users_with_gender AS (
  SELECT
    sbu.*,
    u.gender
  FROM sessions_by_user sbu
  JOIN users u ON u.user_id = sbu.user_id
)

-- select all the information from users_with_gender and perform aggregations
SELECT
  gender,
  COUNT(*) AS n_users,
  SUM(browsing_sessions) AS n_browsings, -- total number of browsing sessions
  SUM(booking_sessions) AS n_bookings,   -- total number of booking sessions
  SUM(cancellation_sessions) AS n_cancels, -- total number of cancellations
  SUM(total_sessions) AS n_sessions, -- total number of sessions
  ROUND(SUM(booking_sessions)::numeric / NULLIF(SUM(total_sessions), 0), 3) AS booking_rate, -- booking rate
  ROUND(SUM(cancellation_sessions)::numeric / NULLIF(SUM(booking_sessions), 0), 3) AS cancel_rate -- cancellation rate
FROM users_with_gender
GROUP BY gender -- group by gender
ORDER BY n_users DESC;


gender n_users n_browsings n_bookings n_cancels n_sessions booking_rate cancel_rate F 54456 223412 125967 3499 352878 0.357 0.028 M 23224 89014 54251 1788 145053 0.374 0.033 O 365 1383 877 28 2288 0.383 0.032

**Note**: Lowering the threshold to more than 5 sessions increases user coverage and improves demographic balance.  
However, the additional users contribute primarily browsing activity, with **limited impact on booking and cancellation rates**.

This suggests that behavioral rates are robust, but signal density per user is reduced.

###### Cohort deifinition: sessions > 3

In [ ]:
%sql
-- select sessions with session_start > '2023-01-04'
WITH filtered_sessions AS (
  SELECT
    user_id,
    trip_id,
    cancellation,
    flight_booked,
    hotel_booked
  FROM sessions
  WHERE session_start > '2023-01-04'
),

-- select eligible users from filtered_sessions
eligible_users AS (
  SELECT user_id
  FROM filtered_sessions
  GROUP BY user_id
  HAVING COUNT(*) > 3  -- users with more than 3 sessions after 2023-01-04
),

-- select cohort users in filtered_sessions
cohort_sessions AS (
  SELECT fs.*
  FROM filtered_sessions fs
  JOIN eligible_users eu ON eu.user_id = fs.user_id
),

-- select all sessions from cohort sessions
sessions_by_user AS (
  SELECT
    user_id,
    COUNT(*) AS total_sessions,
    COUNT(*) FILTER (WHERE trip_id IS NULL) AS browsing_sessions, -- only browsing sessions
    COUNT(*) FILTER (
      WHERE trip_id IS NOT NULL AND cancellation = FALSE
      AND (flight_booked = TRUE OR hotel_booked = TRUE) -- only booking sessions
    ) AS booking_sessions,
    COUNT(*) FILTER (WHERE cancellation = TRUE) AS cancellation_sessions -- only cancellation sessions
  FROM cohort_sessions
  GROUP BY user_id
),

/* select user by gender from filtered sessions_by_user
now to reduce the query execution time */
users_with_gender AS (
  SELECT
    sbu.*,
    u.gender
  FROM sessions_by_user sbu
  JOIN users u ON u.user_id = sbu.user_id
)

-- select all the information from users_with_gender and perform aggregations
SELECT
  gender,
  COUNT(*) AS n_users,
  SUM(browsing_sessions) AS n_browsings, -- total number of browsing sessions
  SUM(booking_sessions) AS n_bookings,   -- total number of booking sessions
  SUM(cancellation_sessions) AS n_cancels, -- total number of cancellations
  SUM(total_sessions) AS n_sessions, -- total number of sessions
  ROUND(SUM(booking_sessions)::numeric / NULLIF(SUM(total_sessions), 0), 3) AS booking_rate, -- booking rate
  ROUND(SUM(cancellation_sessions)::numeric / NULLIF(SUM(booking_sessions), 0), 3) AS cancel_rate -- cancellation rate
FROM users_with_gender
GROUP BY gender -- group by gender
ORDER BY n_users DESC;


gender n_users n_browsings n_bookings n_cancels n_sessions booking_rate cancel_rate F 203043 614589 382842 12058 1009489 0.379 0.031 M 173668 469290 319168 9692 798150 0.400 0.030 O 2571 6946 4758 157 11861 0.401 0.033

**Note**: A threshold of more than 3 sessions substantially increases the user base and yields a **more balanced demographic distribution**.  
At the same time, it introduces a large share of **low-engagement** users, inflating browsing activity without adding meaningful booking signal.

This cohort favors coverage over behavioral depth and is therefore less suitable for segmentation.

##### 5.4.3 Cohort selection rationale

**Conclusion**:
- High thresholds prioritize behavioral depth but reduce demographic balance.
- Low thresholds improve coverage but dilute individual-level signals.
- The selected threshold (>7 sessions) prioritizes behavioral depth while preserving stable booking and cancellation dynamics, making it suitable for segmentation.

The final cohort therefore includes users with more than **7 sessions after 2023-01-04**, providing sufficient behavioral history for downstream user-level aggregation and segmentation.

**Note**: Overall, booking and cancellation rates remain remarkably stable across cohort definitions, indicating that the selected threshold mainly affects user coverage and behavioral depth rather than conversion dynamics.

#### Key Findings

- The database structure is coherent and relationally well-defined, with clean primary keys and no duplicate identifiers across the four core tables.

- Three distinct session types emerge clearly from the data:
browsing-only sessions (`trip_id IS NULL`),
booking sessions (`trip_id NOT NULL AND cancellation = FALSE`),
cancellation sessions (`trip_id NOT NULL AND cancellation = TRUE`).

- Cancellation sessions systematically retain all booking-related flags set to TRUE.
Consequently, the original booking composition must be retrieved from the corresponding booking session when analyzing booking behavior.

- Each cancelled trip consistently consists of exactly two sessions: one booking session followed by one cancellation session.
This confirms a stable and interpretable trip lifecycle at the session level.

- Session duration and page-click distributions reveal a clear separation between short browsing interactions and longer, more engaged booking sessions, supporting their use as meaningful behavioral signals.

- User demographics appear realistic and well balanced, with a strong concentration of working-age travelers and a geographic skew toward North America.

- Flight and hotel activity is heavily concentrated in 2023, justifying the focus on a recent and temporally consistent analysis window aligned with the project objectives.

- Minor data quality issues (e.g. negative hotel nights) are present but isolated and are addressed explicitly in the subsequent data preparation notebook.

- Overall, the analysis confirms that a **semantically consistent and analytically reliable session-level dataset** can be constructed by integrating sessions with user, flight, and hotel data, provided that booking-cancellation temporal relationships are handled explicitly.

These findings provide a validated foundation for user-level aggregation and feature engineering in the subsequent notebooks.

### 6. Cancellations and Temporal Consistency

An important aspect emerging from the exploration of the sessions table is the presence of cancellation sessions whose associated booking sessions may fall outside the selected temporal window.

In particular, some trips are booked before the cohort start date but cancelled after it.
Excluding these booking sessions would result in incomplete trip life cycles and potentially inconsistent interpretations of user behavior.

For this reason, booking sessions associated with cancellations occurring within the cohort window are selectively reintroduced, ensuring a coherent session-level representation of each trip.

**Problem**  
Filtering sessions strictly by date may exclude booking events for trips that are cancelled later,
leading to incomplete or misleading session-level representations.

**Solution**  
Booking sessions associated with cancellations occurring within the cohort window are selectively reintroduced,
ensuring a complete and temporally consistent view of each trip lifecycle.

### 7. Save Session Level Dataset

##### Session-Level Query

Following the guidance from the Head of Marketing, the analysis focuses on users with:
- sessions occurring after 2023-01-04,
- more than 7 recorded sessions.

This cohort definition aims to ensure sufficient behavioral history per user while limiting bias introduced by very recent or inactive users.

Additionally, focusing on recent data reduces potential distortions related to historical effects such as the COVID-19 period.
Given the objectives of customer segmentation and perk allocation, this time-bounded snapshot provides a reasonable and business-aligned representation of active users.

While alternative cohort definitions were considered, this approach was deemed sufficiently robust
for the purpose of customer segmentation and reward design, balancing behavioral depth and recency.

In [ ]:
# ---------------------------------------------------------------
# Final session-level extraction query
#
# This query builds the consolidated session-level dataset used
# in the next notebooks. It applies:
#   - a recency filter (sessions after 2023-01-04),
#   - a minimum activity threshold (>7 sessions per user),
#   - and a temporal consistency rule that reintroduces
#     pre-cutoff booking sessions for trips cancelled later.
#
# The result is a coherent session-level dataset ready for
# cleaning and EDA in Notebook 03.
# ---------------------------------------------------------------

# Run the session-level query and create a Spark DataFrame
session_df = spark.sql("""
WITH
  -- 1. Sessions after cutoff data
  filtered_sessions AS (
    SELECT *
    FROM sessions
    WHERE session_start > '2023-01-04'
  ),

  -- 2. Users with sufficient activity
  filtered_users AS (
    SELECT
      user_id,
      COUNT(*) AS session_count
    FROM filtered_sessions
    GROUP BY user_id
    HAVING COUNT(*) > 7
  ),

  -- 3. Session-level table for filtered users
  session_level AS (
    SELECT
      -- SESSION FIELDS -------------
      s.session_id,
      s.user_id,
      s.trip_id,
      s.session_start,
      s.session_end,
      s.page_clicks,
      s.flight_discount,
      s.flight_discount_amount,
      s.flight_booked,
      s.hotel_discount,
      s.hotel_discount_amount,
      s.hotel_booked,
      s.cancellation,

      -- USERS FIELDS -------------
      u.birthdate,
      u.gender,
      u.married,
      u.has_children,
      u.home_country,
      u.home_city,
      u.home_airport,
      u.home_airport_lat,
      u.home_airport_lon,
      u.sign_up_date,

      -- FLIGHT FIELDS -------------
      f.origin_airport,
      f.destination,
      f.destination_airport,
      f.seats,
      f.return_flight_booked,
      f.departure_time,
      f.return_time,
      f.checked_bags,
      f.trip_airline,
      f.destination_airport_lat,
      f.destination_airport_lon,
      f.base_fare_usd,

      -- HOTEL FIELDS -------------
      h.hotel_name,
      h.nights,
      h.rooms,
      h.check_in_time,
      h.check_out_time,
      h.hotel_per_room_usd

    FROM filtered_sessions s
    INNER JOIN users u ON s.user_id = u.user_id
    LEFT JOIN flights f ON s.trip_id = f.trip_id
    LEFT JOIN hotels h ON s.trip_id = h.trip_id
    WHERE s.user_id IN (SELECT user_id FROM filtered_users)
  ),

  -- 4. Cancellation sessions inside the window
  cancellations AS (
    SELECT *
    FROM session_level
    WHERE cancellation = TRUE
  ),

  -- 5. Retrieve booking sessions BEFORE cutoff for trips cancelled AFTER cutoff
  sessions_trip_booked_before AS (
    SELECT *
    FROM sessions
    WHERE trip_id IN (
      SELECT trip_id
      FROM cancellations
    )
    AND session_start <= '2023-01-04'  -- booking occurred before the cutoff date
  ),

  -- 6.  Build session-level rows for these earlier booking sessions
  session_level_before AS (
    SELECT
      -- SESSION FIELDS -------------
      s.session_id,
      s.user_id,
      s.trip_id,
      s.session_start,
      s.session_end,
      s.page_clicks,
      s.flight_discount,
      s.flight_discount_amount,
      s.flight_booked,
      s.hotel_discount,
      s.hotel_discount_amount,
      s.hotel_booked,
      s.cancellation,

      -- USER FIELDS -------------
      u.birthdate,
      u.gender,
      u.married,
      u.has_children,
      u.home_country,
      u.home_city,
      u.home_airport,
      u.home_airport_lat,
      u.home_airport_lon,
      u.sign_up_date,

      -- FLIGHT FIELDS -------------
      f.origin_airport,
      f.destination,
      f.destination_airport,
      f.seats,
      f.return_flight_booked,
      f.departure_time,
      f.return_time,
      f.checked_bags,
      f.trip_airline,
      f.destination_airport_lat,
      f.destination_airport_lon,
      f.base_fare_usd,

      -- HOTEL FIELDS -------------
      h.hotel_name,
      h.nights,
      h.rooms,
      h.check_in_time,
      h.check_out_time,
      h.hotel_per_room_usd

    FROM sessions_trip_booked_before s
    INNER JOIN users u ON s.user_id = u.user_id
    LEFT JOIN flights f ON s.trip_id = f.trip_id
    LEFT JOIN hotels h ON s.trip_id = h.trip_id
    WHERE s.user_id IN (SELECT user_id FROM filtered_users)
  ),

  -- 7. Combine both parts
  final_tab AS (
    SELECT * FROM session_level
    UNION ALL
    SELECT * FROM session_level_before
  )

SELECT *
FROM final_tab;
""")

In [ ]:
# Convert Spark DataFrame to Pandas DataFrame
session_pdf = session_df.toPandas()

# Quick check of the first few rows
session_pdf.head(5)

session_id  ...      hotel_per_room_usd
0  513347-53b8cae2741a47a38c3a0bd837fc6b06  ...  129.000000000000000000
1  514318-f6964894bb014d4190d4eaac6d49122a  ...  360.000000000000000000
2  544449-9ec11f30602645c587856aaf234f34bd  ...  151.000000000000000000
3  557508-c8e532ef18064eba946c78d2dfde7438  ...  102.000000000000000000
4  559863-83ccf47914b549889033016aa8d81e7b  ...   64.000000000000000000

[5 rows x 41 columns]

In [ ]:
# Quick check
session_pdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49218 entries, 0 to 49217
Data columns (total 41 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   session_id               49218 non-null  object        
 1   user_id                  49218 non-null  int32         
 2   trip_id                  16709 non-null  object        
 3   session_start            49218 non-null  datetime64[ns]
 4   session_end              49218 non-null  datetime64[ns]
 5   page_clicks              49218 non-null  int32         
 6   flight_discount          49218 non-null  bool          
 7   flight_discount_amount   8282 non-null   object        
 8   flight_booked            49218 non-null  bool          
 9   hotel_discount           49218 non-null  bool          
 10  hotel_discount_amount    6206 non-null   object        
 11  hotel_booked             49218 non-null  bool          
 12  cancellation             49218 n

In [ ]:
# Export the DataFrame to CSV
# Save to CSV, overwriting if the file already exists
csv_path = "../data/raw/session_level_raw.csv"
session_pdf.to_csv(csv_path, index=False)

print(f"Session-level CSV saved to: {csv_path}")

Session-level CSV saved to: ../data/raw/session_level_raw.csv


The outcome of this notebook is a session-level dataset that:
- is temporally consistent,
- preserves key behavioral and demographic signals,
- accounts for booking and cancellation dynamics,
- is ready to be reused for further cleaning, aggregation, and exploratory analysis.

The resulting dataset is saved and reused in subsequent notebooks, where user-level aggregation and advanced EDA are performed.

### 8. Summary

- The TravelTide database structure and relationships were analyzed and validated.
- No critical data quality issues were identified, aside from expected NULLs in browsing-only sessions.
- A cohort definition aligned with business objectives was selected.
- A temporally consistent session-level dataset was constructed, accounting for cancellations.
- The resulting dataset is ready for cleaning, feature engineering, and aggregation in Notebook 03.